Python/fileutils.c

```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
     return -1;
 }
 #else   /* MS_WINDOWS */
+int
+_Py_get_blocking(int fd)
+{
+    HANDLE handle;
+    DWORD mode;
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE)
+        return -1;
+
+    if (!GetNamedPipeHandleState(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    return mode == PIPE_WAIT;
+}
+
+int
+_Py_set_blocking(int fd, int blocking)
+{
+    HANDLE handle;
+    DWORD mode;
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE)
+        return -1;
+
+    mode = blocking ? PIPE_WAIT : PIPE_NOWAIT;
+    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
```

Comment:
`PIPE_READMODE_MESSAGE` needs to be retained if it's enabled.

```suggestion
    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
    mode |= blocking ? PIPE_WAIT : PIPE_NOWAIT;
    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
```